sql/core/src/main/scala/org/apache/spark/sql/execution/UnionLoopExec.scala

```diff
@@ -0,0 +1,229 @@
+        true
+      case Project(_, UnionLoopRef(_, _, _)) =>
+        true
+      case _ =>
+        false
+  }
+  }
+  /**
+   * This function executes the plan (optionally with appended limit node) and caches the result,
+   * with the caching mode specified in config.
+   */
+  private def executeAndCacheAndCount(
+                                       plan: LogicalPlan, currentLimit: Int) = {
+    // In case limit is defined, we create a (local) limit node above the plan and execute
+    // the newly created plan.
+    val planOrLimitedPlan = if (globalLimit.isDefined || localLimit.isDefined) {
+      LocalLimit(Literal(currentLimit), plan)
+    } else {
+      plan
+    }
+    val df = Dataset.ofRows(session, planOrLimitedPlan)
+    val newDF = {
+      if (!simpleRecursion) {
+        df.repartition(numPartitions)
+      } else {
+        df
+      }
+    }
+    val count = newDF.count()
+    (newDF, count)
+  }
+
+  /**
+   * In the first iteration, anchor term is executed.
+   * Then, in each following iteration, the UnionLoopRef node is substituted with the plan from the
+   * previous iteration, and such plan is executed.
+   * After every iteration, the dataframe is repartitioned.
+   * The recursion stops when the generated dataframe is empty, or either the limit or
+   * the specified maximum depth from the config is reached.
+   */
+  override protected def doExecute(): RDD[InternalRow] = {
+    val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    val numOutputRows = longMetric("numOutputRows")
+    val numIterations = longMetric("numIterations")
+    val levelLimit = conf.getConf(SQLConf.CTE_RECURSION_LEVEL_LIMIT)
+
+    // currentLimit is initialized from the limit argument, and in each step it is decreased by
+    // the number of rows generated in that step.
+    // If limit is not passed down, currentLimit is set to be zero and won't be considered in the
+    // condition of while loop down (limit.isEmpty will be true).
+    var globalLimitNum = globalLimit.getOrElse(-1)
+    var localLimitNum = localLimit.getOrElse(-1)
+    var currentLimit = Math.max(globalLimitNum, localLimitNum * numPartitions)
+
+    val unionChildren = mutable.ArrayBuffer.empty[LogicalRDD]
+
+    var (prevDF, prevCount) = executeAndCacheAndCount(anchor, currentLimit)
+
+    var currentLevel = 1
+
+    // Main loop for obtaining the result of the recursive query.
+    while (prevCount > 0 && ((globalLimit.isEmpty && localLimit.isEmpty) || currentLimit > 0)) {
+
+      if (levelLimit != -1 && currentLevel > levelLimit) {
+        throw new SparkException(
+          errorClass = "RECURSION_LEVEL_LIMIT_EXCEEDED",
+          messageParameters = Map("levelLimit" -> levelLimit.toString),
+          cause = null)
+      }
+
+      // Inherit stats and constraints from the dataset of the previous iteration.
+      val prevPlan = LogicalRDD.fromDataset(prevDF.queryExecution.toRdd, prevDF, prevDF.isStreaming)
```

Comment:
This is bad for the simple query case. We will recompute the query anyway, why not just use the full query plan as `prevPlan`? The optimizer will combine Projects and Filters.